upgrade-2.x.sh

```diff
@@ -0,0 +1,201 @@
+	    for binary in $tools_binaries; do
+		log "Installing $binary..."
+		curl -s -L -o $tools_path/$binary $download_uri/$binary
+		chmod 755 $tools_path/$binary
+	    done
+	    ;;
+	*)
+	    log ERROR "Binary type $BINARY_TYPE not supported."
+	    ;;
+    esac
+fi
+
+# Find which partition is / and which we should write the update to
+root_part=$(findmnt -n --raw --evaluate --output=source /)
+case $root_part in
+    *p2)
+	root_dev=${root_part%p2}
+	update_part=${root_dev}p3
+	update_part_no=3
+	update_label=resin-rootB
+	;;
+    *p3)
+	root_dev=${root_part%p3}
+	update_part=${root_dev}p2
+	update_part_no=2
+	update_label=resin-rootA
+	;;
+    *)
+	log ERROR "Unknown root partition \"$root_part\"."
+esac
+
+# Stop docker containers
+log "Stopping all containers..."
+systemctl stop update-resin-supervisor.timer > /dev/null 2>&1
+systemctl stop resin-supervisor > /dev/null 2>&1
+$DOCKER stop $($DOCKER ps -a -q) > /dev/null 2>&1 || true
+
+IMAGE=resin/resinos:${TARGET_VERSION}-${SLUG}
+FSARCHIVE=/mnt/data/newos.tar.gz
+
+log "Getting new OS image..."
+# Create container for new version
+CONTAINER=$(docker create ${IMAGE} echo export)
+
+# Export container
+docker export ${CONTAINER} | gzip > ${FSARCHIVE}
+
+# Remove container
+docker rm ${CONTAINER}
+
+# Format alternate root partition
+mkfs.ext4 -F -L ${update_label} ${update_part}
+
+# Mount alternate root partition
+mkdir -p /tmp/updateroot
+mount ${update_part} /tmp/updateroot
+
+# Extract rootfs
+log "Extracting new rootfs..."
+echo quirks >/tmp/root-exclude
+echo resin-boot >>/tmp/root-exclude
+tar -x -X /tmp/root-exclude -C /tmp/updateroot -f ${FSARCHIVE}
+
+# Extract quirks
+tar -x -C /tmp -f ${FSARCHIVE} quirks
+cp -a /tmp/quirks/* /tmp/updateroot/
+rm -rf /tmp/quirks
+
+# XXX hack required until 2.0.7 when a quirk is added for this
+(cd /tmp/updateroot/etc; rm mtab; ln -s ../proc/self/mounts mtab)
+
+# Unmount alternate root partition
+umount /tmp/updateroot
+
+# Extract boot partition, exclude boot_whitelist files
+log "Extracting new boot partition..."
+echo resin-boot/cmdline.txt >/tmp/boot-exclude
+echo resin-boot/config.txt >>/tmp/boot-exclude
+echo resin-boot/splash/resin-logo.png >>/tmp/boot-exclude
+echo resin-boot/uEnv.txt >>/tmp/boot-exclude
+echo resin-boot/EFI/BOOT/grub.cfg >>/tmp/boot-exclude
+# 2.x adds a default config.json, we should avoid clobbering the existing one
+echo resin-boot/config.json >>/tmp/boot-exclude
+tar -x -X /tmp/boot-exclude -C /tmp -f ${FSARCHIVE} resin-boot
+cp -a /tmp/resin-boot/* /mnt/boot/
+
+# Remove OS image
+rm ${FSARCHIVE}
+
+# Switch root partition
+log "Switching root partition..."
+case $SLUG in
+    beaglebone*)
+	echo "resin_root_part=$update_part_no" >/mnt/boot/resinOS_uEnv.txt
+	;;
+    raspberry*)
+	old_root=${root_part#/dev/}
+	new_root=${update_part#/dev/}
+	sed -i -e "s/$old_root/$new_root/" /mnt/boot/cmdline.txt
+	;;
+esac
+
+# Reboot into new OS
+sync
+log "Rebooting into new OS after 5 seconds ..."
+sleep 5
```

Comment:
Mainly to give interested people the chance to Ctrl-C the script if they wish to. ;-) It's probably served it's purpose though and can be removed.